### DISCUSSION TITLE
Pytest problem Lesson 7

### DISCUSSION THREAD

**Discussion by oscarmairey** [smartcontractkit/full-blockchain-solidity-course-py#315] [Q&A] [answered]
Hi, I have a small issue here when running my tets
here is the code:
![image](https://user-images.githubusercontent.com/92814105/140493411-9d1b3b3f-2685-4489-92d4-4fedd9aafc4d.png)
and this code reverts this:
![image](https://user-images.githubusercontent.com/92814105/140493508-778e50c4-4578-4841-aff0-ddaf5a2b3998.png)
whereas the line exceptions should turn the test green 

I've tried everything, even copy pasted Patrick's code directly in my VSCode
Thanks a lot

**Comment by cromewar** [comment]
Hello @Cesarioo 
This happens when you make a change to your contract (typically add a new method) but fail to deploy the contract to the network. so please delete the `build` folder and compile again with `brownie compile`. Also this error comes from the contract, so please verify if you have any typo on `Lottery.sol`, if you want post it in here and I'll help you to check it.

**Reply by oscarmairey** [reply]
Hi Cromewar,
This error is supposed to happen since I wrote that the transaction cannot go through (I used a require) . The problem here is that pytest does not work at supposing the test is true with the exceptions.
with pytest.raises(exceptions.VirtualMachineError) is supposed to let go through testing and making test true when there is a Virtualmachineerror.
Here there is a gas error so the test should not be False but True

**Reply by cromewar** [reply]
Yes you are right the error should trigger a `VirtualMachineError` instead of a Gas error, tha's why I think the error is elsewhere. Did you try to run the teset using `-s` to see what happens step by step? We might use some help from a most experienced eye, @PatrickAlphaC could you give us a hand?

**Reply by PatrickAlphaC** [reply]
Hmm...  Thanks for the flag. What is the command you're using to run this? You could also try `exceptions.ValueError`... maybe brownie updated some codes...

**Reply by ZeroEkkusu** [reply]
Seems to be an issue when testing on local Ganache. By making some intentional errors in code, I can get the other networks to give me `VirtualMachineError`, but would always get `ValueError` with the GUI Ganache (I'm on Lesson 6, btw)
My temporary solution is
```python
e = ValueError if network.show_active() == "ganache-local" else exceptions.VirtualMachineError
with pytest.raises(e):
       #...
```
Let me know if you've found a better one

**Reply by cromewar** [reply]
@Cesarioo try @ZeroEkkusu solution ;)

**Reply by oscarmairey** [reply]
Yes, it works ! Thanks a lot friends !

**Reply by oscarmairey** [reply]
Also an other question: is it normal that my tests take up to 50sec to pass when Patrick's takes only 3sec ?

**Comment by ZeroEkkusu** [comment, marked as answer]
So, the issue is with the GUI Ganache. I've found a few similar issues reported on [the repo](https://github.com/trufflesuite/ganache), as well. Until it gets fixed (or we find what we're doing wrong 😉), you can do
```python
ex = ValueError if network.show_active(
) == "ganache-local" else exceptions.VirtualMachineError
with pytest.raises(ex):
        # ...
```
Another solution is
```python
with pytest.raises(exceptions.VirtualMachineError):
    lottery.enter(
        {"from": get_account(), "value": lottery.getEntranceFee(), "gas_limit": 100000, "allow_revert": True})
```

As for the testing times, this is what I get in Lesson 7:
`brownie test --network development` 00:05
`brownie test --network ganache-local` 00:08
Testing on an actual testnet takes much longer and times may vary by quite a lot

P.S. A better title would be `Getting a ValueError instead of VirtualMachineError when using pytest.raises()`, so that others who face the same problem in the future can find this discussion easily

**Reply by cromewar** [reply]
You are just awesome @ZeroEkkusu.
